src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,671 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Jaroslaw Stelter <jaroslaw.stelter@intel.com>
+//         Pawel Dobrowolski<pawelx.dobrowolski@intel.com>
+
+/*
+ * Dynamic module loading functions.
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component_ext.h>
+#include <sof/common.h>
+#include <sof/ipc/topology.h>
+
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+#include <sof/lib_manager.h>
+#include <sof/audio/module_adapter/module/generic.h>
+
+#include <zephyr/drivers/mm/system_mm.h>
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* 54cf5598-8b29-11ec-a8a3-0242ac120002 */
+
+DECLARE_SOF_UUID("lib_manager", lib_manager_uuid, 0x54cf5598, 0x8b29, 0x11ec,
+		 0xa8, 0xa3, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(lib_manager_tr, SOF_UUID(lib_manager_uuid), LOG_LEVEL_INFO);
+
+/**
+ * DMA buffer
+ */
+struct lib_manager_dma_buf {
+	uintptr_t w_ptr;	/**< write pointer */
+	uintptr_t r_ptr;	/**< read pointer */
+	uintptr_t addr;		/**< buffer start pointer */
+	uintptr_t end_addr;	/**< buffer end pointer */
+	uint32_t size;		/**< buffer size */
+	uint32_t avail;		/**< buffer avail data */
+};
+
+struct lib_manager_dma_ext {
+	struct dma *dma;
+	struct dma_chan_data *chan;
+	struct lib_manager_dma_buf dmabp;
+};
+
+struct ext_library loader_ext_lib;
+
+void *lib_manager_get_config(struct comp_dev *dev)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct module_data *md = &mod->priv;
+	struct module_config *dst = &md->cfg;
+
+	return dst->data;
+}
+
+static int lib_manager_load_data_from_storage(uint8_t *vma, uint8_t *s_addr,
+					      uint32_t size, uint32_t flags)
+{
+	int ret = sys_mm_drv_map_region(vma, POINTER_TO_UINT(NULL),
+					size, SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
```

Comment:
I'm afraid that it was requested in previous reviews. I have no intension in hiding anything.

Reply:
if you mean this comment https://github.com/thesofproject/sof/pull/5796#discussion_r905753274 then it was for an earlier version, the code has changed since then, so, I agree, just `return ret` should be good here and in more cases below